src/dotnet/commands/dotnet-build/ProjectGraphCollector.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Collections.Generic;
+using NuGet.Frameworks;
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.DotNet.ProjectModel.Graph;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.Tools.Build
+{
+    public class ProjectGraphCollector
+    {
+        private readonly Func<string, NuGetFramework, ProjectContext> _projectContextFactory;
+
+        public ProjectGraphCollector(Func<string, NuGetFramework, ProjectContext> projectContextFactory)
+        {
+            _projectContextFactory = projectContextFactory;
+        }
+
+        public IEnumerable<ProjectGraphNode> Collect(IEnumerable<ProjectContext> contexts)
+        {
+            foreach (var context in contexts)
+            {
+                var libraries = context.LibraryManager.GetLibraries();
+                var lookup = libraries.ToDictionary(l => l.Identity.Name);
+                var root = lookup[context.ProjectFile.Name];
+                yield return TraverseProject(root, lookup, context);
+            }
+        }
+
+        private ProjectGraphNode TraverseProject(LibraryDescription library, IDictionary<string, LibraryDescription> lookup, ProjectContext context = null)
+        {
+            var deps = new List<ProjectGraphNode>();
+            foreach (var dependency in library.Dependencies)
+            {
+                var libraryDescription = lookup[dependency.Name];
+
+                if (libraryDescription.Identity.Type.Equals(LibraryType.Project))
+                {
+                    deps.Add(TraverseProject(libraryDescription, lookup));
+                }
+                else
+                {
+                    deps.AddRange(TraversePackage(libraryDescription, lookup));
+                }
+            }
+            var task = context != null ? Task.FromResult(context) : Task.Run(() => _projectContextFactory(library.Path, library.Framework));
+            return new ProjectGraphNode(task, deps, context != null);
+        }
+
+        private IEnumerable<ProjectGraphNode> TraversePackage(LibraryDescription root, IDictionary<string, LibraryDescription> lookup)
```

Comment:
Change type to TargetLibraryWithAssets and rename to libraryWithAssets? There's also MSBuildProjectDescription which should behave just like a package (get ignored but collect sub projects).